File: logcatviewer/src/main/java/com/github/logviewer/MenuClickListener.java
package com.github.logviewer;

import android.content.Context;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;

import com.github.logviewer.databinding.LogcatViewerActivityLogcatBinding;

import java.util.concurrent.Executors;

public class MenuClickListener implements Toolbar.OnMenuItemClickListener {
    private final Context context;
    private final ReadLogcat readLogcat;
    private final LogcatViewerActivityLogcatBinding mBinding;
    private final FloatingWindowLauncher floatingWindowLauncher;

    public interface FloatingWindowLauncher {
        void launchFloatingWindow();
    }

    MenuClickListener(Context context, ReadLogcat readLogcat, LogcatViewerActivityLogcatBinding mBinding, FloatingWindowLauncher floatingWindowLauncher) {
        this.context = context;
        this.readLogcat = readLogcat;
        this.mBinding = mBinding;
        this.floatingWindowLauncher = floatingWindowLauncher;
    }

    @Override
    public boolean onMenuItemClick(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.clear) {
            readLogcat.getAdapter().clear();
        } else if (id == R.id.export) {
            Executors.newSingleThreadExecutor().execute(new ExportLogFileTask(readLogcat.getAdapter().getData(), mBinding, context));
        } else if (id == R.id.floating) {
            floatingWindowLauncher.launchFloatingWindow();
        } else if (id == R.id.Verbose || id == R.id.Debug || id == R.id.Info || id == R.id.Warning || id == R.id.Error || id == R.id.Fatal) {
            readLogcat.getAdapter().getFilter().filter(item.getTitle());
        } else {
            return false;
        }

        return true;
    }
}

File: logcatviewer/src/main/kotlin/com/github/logviewer/MenuClickListener.kt
package com.github.logviewer

import android.content.Context
import android.content.Intent
import android.view.MenuItem
import androidx.appcompat.widget.Toolbar
import androidx.core.content.FileProvider
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.io.File
import java.io.FileOutputStream

class MenuClickListener internal constructor(
    private val context: Context,
    private val readLogcat: ReadLogcat,
    private val floatingWindowLauncher: () -> Unit = {},
) : Toolbar.OnMenuItemClickListener {
    @OptIn(DelicateCoroutinesApi::class)
    override fun onMenuItemClick(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.clear -> readLogcat.clear()
            R.id.export -> exportLogFile()
            R.id.floating -> floatingWindowLauncher()
            R.id.Verbose, R.id.Debug, R.id.Info, R.id.Warning, R.id.Error, R.id.Fatal ->
                readLogcat.filter(item.title)
            else -> return false
        }
        return true
    }

    @DelicateCoroutinesApi
    fun exportLogFile() {
        GlobalScope.launch(Dispatchers.IO) {
            File(context.externalCacheDir, "yuhaiin.log").apply {
                writeText("Yuhaiin Logcat:\n")
                Runtime.getRuntime()
                    .exec(arrayOf("logcat", "-d")).inputStream.use { input ->
                        FileOutputStream(this, true).use {
                            input.copyTo(it)
                        }
                    }

                val authority = "${context.packageName}.logcat_fileprovider"
                val uri = FileProvider.getUriForFile(context, authority, this)
                context.startActivity(
                    Intent.createChooser(
                        Intent(Intent.ACTION_SEND)
                            .setType("text/plain")
                            .setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                            .putExtra(Intent.EXTRA_STREAM, uri),
                        "Export Logcat"
                    ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                )
            }
        }
    }
}